effectful/ops/syntax.py

```diff
@@ -597,6 +597,66 @@ def kwargs(self):
         return self._kwargs
 
 
+def register_dataclass(cls: type):
+    """
+    Registers a dataclass to effectful.
+
+    """
+    name = cls.__name__
+
+    term_name = f"_{name}Term"
+    bases = (Term, cls)
+    term_cls = _DataclassTermMeta(term_name, bases, {})
+
+    defdata.register(cls)(term_cls)
+    return cls
+
+
+class _DataclassTermMeta(type(_BaseTerm)):  # type: ignore
+    def __new__(mcls, name, bases, ns):
+        base_dt = None
+        for b in bases:
+            if dataclasses.is_dataclass(b):
+                base_dt = b
+                break
+
+        if base_dt is None:
+            raise TypeError("No dataclass base found")
+
+        for f in dataclasses.fields(base_dt):
+            attr = f.name
+            field_type = f.type
+
+            def make_getter(a, return_type):
+                def getter(self) -> return_type:
+                    if isinstance(self, Term):
+                        raise NotHandled
+                    return self.__dict__[a]
```

Comment:
Can we use `functools.wraps` around the original property to give this the appropriate name and type annotations?

Reply:
@jfeser after looking into it, dataclass fields aren't properties so `functools.wraps` doesn't work here - there isn't an original property to wrap.